core/src/main/java/com/linecorp/armeria/internal/AnnotatedHttpServiceParamUtil.java

```diff
@@ -208,6 +216,127 @@ public static Object convertParameter(@Nullable String value,
+            return value;
+        }
+        return getName(nameRetrievalTarget);
+    }
+
+    /**
+     * Returns the value of the {@link Header} annotation which is specified on the {@code element} if
+     * the value is not blank. If the value is blank, it returns the name of the specified
+     * {@code element} object which is an instance of {@link Parameter} or {@link Field}.
+     *
+     * <p>Note that the name of the specified {@code element} will be converted as
+     * {@link CaseFormat#LOWER_HYPHEN} that the string elements are separated with one hyphen({@code -})
+     * character. The value of the {@link Header} annotation will not be converted because it is clearly
+     * specified by a user.
+     */
+    @Nullable
+    public static String findHeaderName(AnnotatedElement element) {
+        return findHeaderName(element, element);
+    }
+
+    /**
+     * Returns the value of the {@link Header} annotation which is specified on the {@code element} if
+     * the value is not blank. If the value is blank, it returns the name of the specified
+     * {@code nameRetrievalTarget} object which is an instance of {@link Parameter} or {@link Field}.
+     *
+     * <p>Note that the name of the specified {@code nameRetrievalTarget} will be converted as
+     * {@link CaseFormat#LOWER_HYPHEN} that the string elements are separated with one hyphen({@code -})
+     * character. The value of the {@link Header} annotation will not be converted because it is clearly
+     * specified by a user.
+     */
+    @Nullable
+    public static String findHeaderName(AnnotatedElement element, Object nameRetrievalTarget) {
+        requireNonNull(element, "element");
+        requireNonNull(nameRetrievalTarget, "nameRetrievalTarget");
+        final Header header = element.getAnnotation(Header.class);
+        if (header == null) {
+            return null;
+        }
+
+        final String value = header.value();
+        if (DefaultValues.isSpecified(value)) {
+            checkArgument(!value.trim().isEmpty(), "value: %s (expected: not blank)", value);
+            return value;
+        }
+        return toHeaderName(getName(nameRetrievalTarget));
+    }
+
+    /**
+     * Returns whether the specified {@code element} is annotated with a {@link Param} or {@link Header}.
+     */
+    public static boolean hasParamOrHeader(AnnotatedElement element) {
+        return element.isAnnotationPresent(Param.class) ||
+               element.isAnnotationPresent(Header.class);
+    }
+
+    private static String getName(Object element) {
+        if (element instanceof Parameter) {
+            final Parameter parameter = (Parameter) element;
+            checkArgument(parameter.isNamePresent(),
+                          "cannot obtain the name of the parameter or field automatically. " +
+                          "Please check whether your compile option includes '-parameters'. " +
```

Comment:
check whether ... -> make sure you compiled your code with '-parameters' option. ?